### PR TITLE
Custom fusion

### PR DESCRIPTION
**Pull request by ArthurBrussee**
### Related Issues/PRs

https://github.com/tracel-ai/burn/issues/1970

### Changes
Expose enough information to allow custom operations to use fusing. Also expose a function to resolve a tensor from a fusion backend.

Draft until https://github.com/tracel-ai/cubecl/pull/256 lands & some other bits are cleaned up.

This also fixes https://github.com/tracel-ai/burn/issues/1970 by bailing on fusing when going over max bindings.

### CODE REVIEW

File: crates/burn-fusion/src/client/base.rs [hunk 1]
```diff
@@ -67,6 +67,18 @@ where
...
+    where
+        B: FusionBackend<FusionRuntime = R>;
+    /// Change the client of the given float tensor.
+    fn resolve_tensor_int<B>(&self, tensor: FusionTensor<R>) -> B::IntTensorPrimitive
+    where
+        B: FusionBackend<FusionRuntime = R>;
+    /// Change the client of the given float tensor.
+    fn resolve_tensor_bool<B>(&self, tensor: FusionTensor<R>) -> B::BoolTensorPrimitive
+    where
+        B: FusionBackend<FusionRuntime = R>;
```
Comment (nathanielsimard) [inline crates/burn-fusion/src/client/base.rs:80]: Docs is not up to date.

File: crates/burn-jit/src/fusion/on_write/builder.rs [hunk 2]
```diff
@@ -16,10 +16,40 @@ use cubecl::ir::Elem;
...
+    }
+
+    fn register(&mut self, build: impl FnOnce(&mut FuseOnWriteTraceBuilder)) -> bool {
+        let mut cloned = self.builder.clone();
+        build(&mut cloned);
+        if cloned.estimate_bindings() > self.max_bindings {
+            return false;
+        }
+        self.builder = cloned;
+        true
```
Comment (nathanielsimard) [inline crates/burn-jit/src/fusion/on_write/builder.rs:47]: cloning every time is a bit wasteful. I would only do it if `builder_current_state.estimate_bindings() + new_op.nodes() > self.max_bindings`

File: crates/burn-jit/src/fusion/on_write/builder.rs [hunk 3]
```diff
@@ -28,6 +58,13 @@ impl OptimizationBuilder<FuseOnWriteTrace> for FuseOnWriteBuilder {
             return;
         }
 
+        // TODO: If tensors are re-used, we should count them only once.
+        self.num_tensors += op.nodes().len();
+        if self.num_tensors >= 10 {
+            self.status = OptimizationStatus::Closed;
+            return;
+        }
```
Comment (nathanielsimard) [inline crates/burn-jit/src/fusion/on_write/builder.rs:66]: It should use `estimate_bindings` I think, and 10 should be replaced by the value in the device's hardware properties.

File: crates/burn-tensor/src/repr/operation.rs [hunk 4]
```diff
@@ -16,6 +16,21 @@ use crate::{
 
 use super::{QuantizationParametersDescription, QuantizedTensorDescription};
 
+/// Custom operation in fusion stream, declaring it's inputs and outputs.
+#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
+pub struct CustomOpDescription {
+    /// Input tensors used in this the custom operation.
```
Comment (nathanielsimard) [inline crates/burn-tensor/src/repr/operation.rs:22]: I would add an `id` field (a string for now). And maybe a `scalars` field too, but it's not important to have it now.

File: crates/burn-router/src/runner.rs [hunk 5]
```diff
@@ -1202,6 +1202,9 @@ where
                     handles.register_float_tensor::<B>(&desc.out.id, output);
                 }
             },
+            OperationDescription::Custom(_) => {
+                panic!("Can't execute custom operation here")
```
Comment (nathanielsimard) [inline crates/burn-router/src/runner.rs:1206]: At some point we should resolve a function based on the custom description id and call it with the handles and the description as input.

File: crates/burn-jit/src/fusion/on_write/builder.rs (hunk 2 above)
Comment (ArthurBrussee) [reply crates/burn-jit/src/fusion/on_write/builder.rs]: That doesn't really work - we don't know how many nodes someone is adding in the build function. And we can't easily keep track of it as adding them mutates the builder.

estimate_bindings() also seems more expensive than a clone in the first place?

I've at least added a fast path for the first operation which should be common. If that's not enough might need to rethink how these operations are added so we can count them more easily.

File: crates/burn-jit/src/fusion/on_write/builder.rs (hunk 3 above)
Comment (ArthurBrussee) [reply crates/burn-jit/src/fusion/on_write/builder.rs]: Woops this was leftover dummy stuff

**Comment by ArthurBrussee** [comment]
I've also added a fix for a deadlock in fusion on wasm in this PR now. It might or might not make things run a bit smoothly on other platforms too, as it was incorrectly holding the server lock while waiting for a read.

**Review by nathanielsimard — APPROVED** [review]

**Pull request merged**